### ISSUE TITLE
Code style improvements

### ISSUE THREAD

**Issue by wetneb** [OpenRefine/OpenRefine#2338]
We have not been enforcing a consistent code style lately. This causes some issues in inconsistent indentation, diverging name conventions, etc.

- [x] decide which style code should follow (probably broken down by languages)
- [x] convert the existing code to match these guidelines
- [x] set up continuous integration to enforce these styles

This would need to be done both for Java and Javascript.

**Comment by thadguidry** [comment]
"I think", we have some prior discussion on mailing lists...or even old issues, that had some of this discussion (Tom and Stefano had preferences)...but it's 2020 now...so best to start fresh and make decisions on current practices. (just thinking about ECMA and all those debates they themselves had back in the day, hahaha)

**Comment by thadguidry** [comment]
Actually, yeah, we had some decisions and some of those landed in the /IDE path

**Comment by kushthedude** [comment]
@wetneb @thadguidry What about implementing ESLint for JS and Checkstyle for Java >

**Comment by thadguidry** [comment]
@kushthedude yeah probably Checkstyle for Java...doesn't SonarQube and IntelliJ both use that as a default now?  Can you check?

**Comment by kushthedude** [comment]
Can't checkstyle also enforce the coding styles such as Linting, White-Spaces and other Google Java Style Standard Rules

**Comment by kushthedude** [comment]
Also, ESLint can do the same for the js code with an option to add tests into the CI.

**Comment by kushthedude** [comment]
@thadguidry Shall I go ahead implementing the same ?

**Comment by thadguidry** [comment]
There's a style guide and then there's enforcing compliance of the style.
It sounds like you want to make the enforcement and IDE setup for developers easier?
Optionally enforcing checks on our CI/CD for pull requests?

You can look to a few Apache projects and GitHub projects to see how they do this (only requires us to supply a few extra files historically, but today, I think Eclipse and IntelliJ ship with a few default styles like Google Java Style ...check into this yourself to see. )  http://help.eclipse.org/2019-12/index.jsp
Here's the official location of Google Style Guides for reference (but tools already grab from here): https://github.com/google/styleguide
Eclipse Help setup: http://help.eclipse.org/2019-12/index.jsp

And here's a snippet of instructions that we could also borrow and put somewhere in our docs, perhaps in the CONTRIBUTOR.md or README.md  (I don't know...check on how Apache projects set this up and give developers the info, I would say, even though we are not an Apache project, historically we have pulled in their best practices)
https://github.com/HPI-Information-Systems/Metanome/wiki/Installing-the-google-styleguide-settings-in-intellij-and-eclipse#installing-the-coding-style-settings-in-eclipse

So I would say...
1. If we can have a clean way for developers to clone OpenRefine and have their IDE like IntelliJ or Eclipse already setup for Google Code Style, that would be awesome.
2. As well as provide instructions in the README.md or CONTRIBUTOR.md on how to setup the Google Code Style (maybe just links to .
3. For CI/CD, we should look to enable Checkstyle, through GitHub Actions, Travis, and AppVeyor with Maven support.

**Comment by kushthedude** [comment]
I am planning for developers environment with ci/cd check too.

**Comment by thadguidry** [comment]
I like a lot of the info that Apache Spark pulls together on their Contributing doc (Code Style is at the bottom of page):  http://spark.apache.org/contributing.html

**Comment by kushthedude** [comment]
I will check it out.

**Comment by kushthedude** [comment]
@wetneb @thadguidry I set up the style enforcements for JS and have used standard js code style using ESLint and I have also set up npm and package.lock for our project. Should I push the changes?

**Comment by thadguidry** [comment]
@kushthedude Sure, submit with a PR so that we can review as a community.

**Comment by thadguidry** [comment]
Hi @kushthedude We are awaiting your new PR for the Javascript style.  (Then we can work on the CI for the enforcement for that in a separate PR).  We want to follow [the task steps](https://github.com/OpenRefine/OpenRefine/issues/2338#issue-572314326) in the top of this issue.

**Comment by kushthedude** [comment]
To enforce ESlint there will be a need for node engine. As webjars can't utilise the rules for the ESLint standard js style. Hence first we need to decide do we need to go for yarn/npm or any other alternative? Then we can go for javascript styling.

**Comment by wetneb** [comment]
@thadguidry I don't think there is any particular rush to get this done - and we should not pressure GSoC/Outreachy applicants to work on more things before the internships start :)

**Comment by kushthedude** [comment]
@wetneb @thadguidry Do we have to move with Checkstyle for Java ?

**Comment by thadguidry** [comment]
@kushthedude Antonin as our lead and having to deal with the Checkstyle...and fallout from it... will make that final determination.

**Comment by tfmorris** [comment]
> "I think", we have some prior discussion on mailing lists...or even old issues, that had some of this discussion (Tom and Stefano had preferences)...but it's 2020 now...so best to start fresh and make decisions on current practices. (just thinking about ECMA and all those debates they themselves had back in the day, hahaha)

I think that unless there's a strong reason not to, which should stick with the conventions that were used to write the bulk of the code. Anything else adds a ton of noise.

The conventions were documented in the form of Eclipse project settings that defined things like 4 space indentation, import ordering, etc for Java.  For Javascript, I'm not sure we did much more than specify 2 space indents. These got deleted as part of the move to Maven, but I've relocated them and just need to separate out the style preferences from the others.

The biggest issue that I see with new code is the use of tabs for indentation (pure tabs, I think, as opposed to mixed spaces & tabs) instead of the historical 4 spaces. This most likely means that people have their tabs stops set at 4 spaces too, instead of the usual 8, so that things align visually for them.

I have mixed feelings about doing a mass cleanup. This tends to introduce a lot of noise that makes it more difficult to use commit histories and git blame. Whether we do cleanups all at once or incrementally, they should be in separate commits to make it easy to differentiate functional changes and formatting changes.

**Comment by wetneb** [comment]
It's actually quite easy to enforce code style settings defined in the format Eclipse accepts, using this Maven plugin:
```
       <plugin>
        <groupId>net.revelc.code.formatter</groupId>
        <artifactId>formatter-maven-plugin</artifactId>
        <version>2.17.0</version>
      </plugin>
```

This reformats Java files at build time.
I am tempted to integrate this, but would ideally wait for 4.0 to become the master branch, because I am worried about spurious merge conflicts generated by this if we reformat everything. But maybe it is fine.

**Comment by wetneb** [comment]
Note that we should also add some configuration settings in our Maven configuration so that Eclipse picks up the style configuration file:
https://maven.apache.org/plugins/maven-eclipse-plugin/examples/load-code-styles.html

**Comment by wetneb** [comment]
Nice setup from Gephi: https://github.com/gephi/gephi/issues/2337#event-5706587778

**Comment by wetneb** [comment]
I have made some progress on this:
First I have adapted the existing Eclipse style file to better match the current state of the source code, so that applying it makes minimal changes to the code base.
Then I have run linting both on the master and 4.0 branches, merging master into 4.0 before and after.

I have made a small test which consisted in editing a part that was affected by the linting but was common to both branches: it got merged without a problem. So I am hopeful that this linting will not create too many merge conflicts when merging changes from master into 4.0.
